Add the second tutorial page. (#234)
square/workflow-kotlin: samples/tutorial/Tutorial2.md

# Step 2

_Multiple Screens and Navigation_

## Setup

To follow this tutorial, launch Android Studio and open this folder (`samples/tutorial`).

Start from the implementation of `tutorial-1-complete` if you're skipping ahead.

## Second Workflow

Let's add a second screen and workflow so we have somewhere to land after we log in. Our next screen will be a list of "todo" items, as todo apps are the best apps.

Create a new screen/`LayoutRunner` pair called `TodoList`:

Add the provided `TodoListViewBinding` from `tutorial-views` as a subview to the newly created view controller:

```kotlin
/**
 * This should contain all data to display in the UI.
 *
 * It should also contain callbacks for any UI events, for example:
 * `val onButtonTapped: () -> Unit`.
 */
data class TodoListScreen()

class TodoListLayoutRunner(

>   private val todoListBinding: TodoListViewBinding

Reviewer: ```suggestion
  /** From `todo_list_view.xml`. */
  private val todoListBinding: TodoListViewBinding
```

WDYT?

```kotlin
object TodoListWorkflow : StatefulWorkflow<Unit, State, Nothing, TodoListScreen>() {

  data class State()

  override fun initialState(
    props: Unit,
    snapshot: Snapshot?
  ) = Unit

  override fun render(
    props: Unit,
    state: State,
    context: RenderContext
  ): TodoListScreen {
    return TodoListScreen(
      name = "",
      todoTitles = emptyList(),
      onTodoSelected = {}
      onBack = {}
    )
  }

  override fun snapshotState(state: State): Snapshot? = null
}
```

### Showing the new screen and workflow

For now, let's just show this new screen instead of the login screen/workflow. Update the activity to show the `TodoListWorkflow`:

```kotlin
private val viewRegistry = ViewRegistry(
    WelcomeLayoutRunner,
    TodoListLayoutRunner
)

class TutorialActivity : AppCompatActivity() {

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    setContentWorkflow(viewRegistry) {
      Config(TodoListWorkflow, Unit)
    }
  }
}
```

Run the app again, and now the empty todo list (table view) will be shown:

![Empty Todo List](images/empty-todolist.png)

## Populating the Todo List

> The empty list is rather boring, so let's fill it in with some sample data for now. Update the `State` type to include a list of todo model objects and change `initialState` to include a default one:

Reviewer: I keep hearing @davidapgar read this. Where are the cat pics?

      text = resources.getString(R.string.todo_list_welcome, rendering.name)
    }

    adapter.todoList = rendering.todoTitles
    adapter.notifyDataSetChanged()
  }

  // …
}
```

Finally, update `render` for `TodoListWorkflow` to send the titles of the todo models whenever the screen is updated:

```kotlin
object TodoListWorkflow : StatefulWorkflow<Unit, State, Nothing, TodoListScreen>() {

  override fun render(
    props: Unit,
    state: State,
    context: RenderContext
  ): TodoListScreen {
    val titles = state.todos.map { it.title }
    return TodoListScreen(
      name = "",
      todoTitles = titles,
      onTodoSelected = {},
      onBack = {}
    )
  }

  // …
}
```

Run the app again, and now there should be a single visible item in the list:

![Todo list hard coded](images/tut2-todolist-example.png)

## Composition and Navigation

Now that there are two different screens, we can make our first workflow showing composition with a single parent and two child workflows. Our `WelcomeWorkflow` and `TodoListWorkflow` will be the leaf nodes with a new workflow as the root.

### Root Workflow

Create a new workflow called `Root` with the templates.

> We'll start with the `RootWorkflow` returning only showing the `WelcomeScreen` via the `WelcomeWorkflow`. Updated the `Rendering` type and `render` to have the `RootWorkflow` defer to a child:

Reviewer: ```suggestion
We'll start with the `RootWorkflow` returning only showing the `WelcomeScreen` via the `WelcomeWorkflow`. Update the `Rendering` type and `render` to have the `RootWorkflow` defer to a child:
```

```kotlin
object RootWorkflow : StatefulWorkflow<Unit, Unit, Nothing, WelcomeScreen>() {

  override fun initialState(
    props: Unit,
    snapshot: Snapshot?
  ): Unit = Unit

  override fun render(
    props: Unit,
    state: Unit,
    context: RenderContext
  ): Any {
    // Render a child workflow of type WelcomeWorkflow. When renderChild is called, the

>     // infrastructure will create a child workflow with state if one is not already running.

Reviewer: ```suggestion
    // infrastructure will start a child workflow session with state, if one is not already running.
```

```kotlin
object TodoListWorkflow : StatefulWorkflow<Unit, State, Nothing, TodoListScreen>() {
```

And update the `render` method of the `RootWorkflow`:

```kotlin
object RootWorkflow : StatefulWorkflow<Unit, State, Nothing, Any>() {

  // …

  override fun render(
    props: Unit,
    state: State,
    context: RenderContext
  ): Any {
    when (state) {
      // When the state is Welcome, defer to the WelcomeWorkflow.
      is Welcome -> {
        // Render a child workflow of type WelcomeWorkflow. When renderChild is called, the
        // infrastructure will create a child workflow with state if one is not already running.
        val welcomeScreen = context.renderChild(WelcomeWorkflow) { output ->
          // When WelcomeWorkflow emits LoggedIn, turn it into our login action.
          login(output.name)
        }
        return welcomeScreen
      }

      // When the state is Todo, defer to the TodoListWorkflow.
      is Todo -> {
        val todoScreen = context.renderChild(TodoListWorkflow, props = ListProps(state.name)) {
          logout()
        }
        return todoScreen
      }
    }
  }

  // …
}
```

> ### Props (TODO better title?)

Author: Went with "Workflow Props" to match the "Workflow Output" section above.

So far we are gathering a name from the welcome screen, and there's a place on the todo list screen to display the name, but we're not passing the name from the welcome screen to the next one. We'll fix this by passing the name down from the `RootWorkflow` to the `TodoListWorkflow` via "props".

> TK

Reviewer: TK

```kotlin
object RootWorkflow : StatefulWorkflow<Unit, State, Nothing, BackStackScreen<Any>>() {

  // …

  @OptIn(WorkflowUiExperimentalApi::class)
  override fun render(
    props: Unit,
    state: State,
    context: RenderContext
  ): BackStackScreen<Any> {

    // Our list of back stack items. Will always include the "WelcomeScreen".
    val backstackScreens = mutableListOf<Any>()

    // Render a child workflow of type WelcomeWorkflow. When renderChild is called, the
    // infrastructure will create a child workflow with state if one is not already running.
    val welcomeScreen = context.renderChild(WelcomeWorkflow) { output ->
      // When WelcomeWorkflow emits LoggedIn, turn it into our login action.
      login(output.name)
    }
    backstackScreens += welcomeScreen

    when (state) {
      // When the state is Welcome, defer to the WelcomeWorkflow.
      is Welcome -> {
        // We always add the welcome screen to the backstack, so this is a no op.
      }

      // When the state is Todo, defer to the TodoListWorkflow.
      is Todo -> {

>         backstackScreens.addAll(context.renderChild(TodoListWorkflow, props = ListProps(state.name)) {

Reviewer: Confused by `addAll` and `.frames` here, wondering if you skipped a step. It looks like this code assumes that `TodoListWorkflow` also renders `BackStackScreen`, but we haven't made such a change.

Author: I refactored some code, and forgot to refactor the tutorial. Fixed.

- `initialState` – the first time the parent renders the child, the props value it provides is passed to this function.
- `onPropsChanged` – this function is only called when the parent passes a different props value to `renderChild` than it did from the last time it called `renderChild`. This method gets both the old and the new props, and can return an updated state to reflect those changes.
- `render` – gets the props the parent passed to `renderChild`. The first time this workflow is rendered, this props will be the same as the value that was passed to `initialState`.
- `WorkflowAction.apply` – when an action is applied, the `Updater` receiver has a reference to the last props used to render the workflow that the action is being applied to. In terms of sequencing, actions always happen _between_ renders.

> A workflow's props is similar to its state in a sense: any time the workflow is "alive", it has a current state and a current props. The _state_ is readable and writable, and owned by the workflow – the workflow changes its own state. The _props_ are read-only, and owned by the workflow's parent. The child can't change its props, it can only observe the props its parent decided to pass down.

Reviewer: I really like the read-write, read-only distinction.

> Note that unlike output, which is how a child sends events to its parent, props does not represent events. In fact, another way to think of props is another kind of state – the "public" part of its state, if you will.

Reviewer: Perfect. This is great.
